frame/nft-fractionalisation/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+
+		type AssetBalance: AtLeast32BitUnsigned
+			+ codec::FullCodec
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ sp_std::fmt::Debug
+			+ Default
+			+ From<u64>
+			+ IntegerSquareRoot
+			+ Zero
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		type AssetId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ codec::HasCompact
+			+ From<u32>
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ PartialOrd
+			+ TypeInfo;
+
+		type Assets: Inspect<Self::AccountId, AssetId = Self::AssetId, Balance = Self::AssetBalance>
+			+ Create<Self::AccountId>
+			+ InspectEnumerable<Self::AccountId>
+			+ Mutate<Self::AccountId>
+			+ MutateMetadata<Self::AccountId>
+			+ Transfer<Self::AccountId>;
+
+		type Items: NonFungiblesInspect<
+				Self::AccountId,
+				ItemId = Self::ItemId,
+				CollectionId = Self::CollectionId,
+			> + NonFungiblesTransfer<Self::AccountId>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		#[pallet::constant]
+		type BuybackThreshold: Get<u32>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn assets_minted)]
+	// TODO: query amount minted from pallet assets instead of storing it locally.
+	// Add a public getter function to pallet assets.
+	pub type AssetsMinted<T: Config> =
+		StorageMap<_, Twox64Concat, AssetIdOf<T>, AssetBalanceOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn asset_to_nft)]
+	// TODO: store information about Asset ID and the corresponding Collection and Item ID.
+	pub type AssetToNft<T: Config> =
+		StorageMap<_, Twox64Concat, AssetIdOf<T>, (T::CollectionId, T::ItemId), OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		NFTLocked(T::CollectionId, T::ItemId),
+		NFTUnlocked(T::CollectionId, T::ItemId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetAlreadyRegistered,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(2).ref_time())]
+		/// Pallet's account must be funded before lock is possible!
+		/// 5EYCAe5gjC5dxKPbV2GPQUetETjFNSYZsSwSurVTTXidSLbh
+		pub fn lock_nft_create_asset(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+			item_id: T::ItemId,
+			asset_id: AssetIdOf<T>,
+			beneficiary: T::AccountId,
+			min_balance: AssetBalanceOf<T>,
+			amount: AssetBalanceOf<T>,
+		) -> DispatchResult {
+			let _who = ensure_signed(origin.clone())?;
+			let admin_account_id = Self::pallet_account_id();
+
+			Self::do_lock_nft(collection_id, item_id)?;
+			Self::do_create_asset(asset_id, admin_account_id, min_balance)?;
+			Self::do_mint_asset(asset_id, &beneficiary, amount)?;
+
+			// Mutate this storage item to retain information about the amount minted.
+			<AssetsMinted<T>>::try_mutate(
+				asset_id,
+				|assets_minted| -> Result<(), DispatchError> {
+					match assets_minted.is_some() {
+						true =>
+							*assets_minted = Some(assets_minted.unwrap().saturating_add(amount)),
```

Comment:
could you please explain why are you using the `unwrap()` here? i would try to avoid using `unwrap`s and something that handles the error as `unwrap_or` for example.

Reply:
The `unwrap()` happens only if `assets_minted.is_some()` is `true`

Reply:
i see what you mean, there is no way that the unwrap will throw an error bacause of the `is_some()`, i wonder if there is a way to simplify the code without the match.